Nullptr crash in MediaRecorder::requestDataInternal()

pratiksha-99 opened the pull request:
#### 1d9dbbef71331e403a6c691d698496672218816e
<pre>
Nullptr crash in MediaRecorder::requestDataInternal()
<a href="https://bugs.webkit.org/show_bug.cgi?id=288989">https://bugs.webkit.org/show_bug.cgi?id=288989</a>
<a href="https://rdar.apple.com/144407606">rdar://144407606</a>

Reviewed by Jean-Yves Avenard.

Added a nullptr check for buffer object. As per W3C spec, the blob can be empty if no data is gathered.
We do not have reliable test script to repro this edge case race condition.

* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::requestDataInternal):

Canonical link: <a href="https://commits.webkit.org/291974@main">https://commits.webkit.org/291974@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d9b074759a0d517f5c0a001e6f1575261a7c173

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13672 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3417 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99096 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44612 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22102 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71780 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29125 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97087 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10366 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84957 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52121 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10050 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2630 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43928 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80296 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2717 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101135 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21137 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15395 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80788 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21389 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80956 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80163 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24711 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2069 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/14299 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15178 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21121 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26300 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20808 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24268 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22549 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
```diff
@@ -255,6 +255,10 @@ ExceptionOr<void> MediaRecorder::requestDataInternal(ReturnDataIfEmpty returnDat
 
+        //  As per W3C spec, if no data has been captured, the MediaRecorder will still fire the dataavailable event, but the Blob associated with that event will be empty.
+        if (!buffer)
+            dispatchEvent(createDataAvailableEvent(scriptExecutionContext(), { }, { }, 0));
```
that may provide a timecode that goes back in time, timecode must be monotonically increasing.

pratiksha-99 replied:
Updated to use the current timecode.

rniwa commented on Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
```diff
@@ -255,6 +255,10 @@ ExceptionOr<void> MediaRecorder::requestDataInternal(ReturnDataIfEmpty returnDat
 
+        //  As per W3C spec, if no data has been captured, the MediaRecorder will still fire the dataavailable event, but the Blob associated with that event will be empty.
+        if (!buffer)
+            dispatchEvent(createDataAvailableEvent(scriptExecutionContext(), { }, { }, timeCode));
```
Don't we need to return here to avoid the crash??

rniwa commented on LayoutTests/fast/mediastream/mediarecorder-invalid-buffer-crash.html:
```diff
@@ -0,0 +1,76 @@
+<script>
```
This test doesn't reproduce the crash.

pratiksha-99 replied:
Reproduces sometimes so most likely this is race created by this fuzzing test case which can make the buffer empty as MediaRecorder is initialized with a stream that doesn’t have any media tracks + its start/stop puts in an unexpected state. I dont have much context to make the test reliable.

rniwa replied:
Okay, in that case, it might be better not to add a test. It's rather misleading to add a test that sometimes reproduces the issue.

pratiksha-99 committed:
Nullptr crash in MediaRecorder::requestDataInternal()
https://bugs.webkit.org/show_bug.cgi?id=288989
rdar://144407606

Reviewed by Jean-Yves Avenard.

Added a nullptr check for buffer object. As per W3C spec, the blob can be empty if no data is gathered.
We do not have reliable test script to repro this edge case race condition.

* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::requestDataInternal):

Canonical link: https://commits.webkit.org/291974@main
